docs/build_system.md

```diff
@@ -0,0 +1,70 @@
+5. The `Rakefile` `:compile` task ensures the above prerequisites are done, then calls `make`,
+  and uses `Rake::ExtensionTask` to compile the C extension (using its `extconf.rb`), which uses `librubyparser.a`
+
+This way there is minimal duplication, and each level builds on the previous one.
+
+## The various ways to build YARP
+
+### Building from ruby/yarp repository with `bundle exec rake`
+
+`rake` calls `make` and then uses `Rake::ExtensionTask` to compile the C extension (see above).
+
+### Building the yarp gem by `gem install/bundle install`
+
+The gem contains the pre-generated templates, as well as `configure` and `config.h.in`
+When installing the gem, `extconf.rb` is used and that:
+* runs `./configure` which creates the `Makefile` and `include/yarp/config.h`
+* runs `make build/librubyparser.a`
+* compiles the C extension with mkmf
+
+When installing the gem on JRuby and TruffleRuby, no C extension is built, so instead of the last step,
+there is Ruby code using Fiddle which uses `librubyparser.so/dylib/dll`
+to implement the same methods as the C extension, but using serialization instead of many native calls/accesses
+(JRuby does not support C extensions, serialization is faster on TruffleRuby than the C extension).
+
+### Building the yarp gem from git, e.g. `gem 'yarp', github: 'ruby/yarp'`
+
+The same as above, except the `extconf.rb` additionally runs first:
+* `templates/template.rb` to generate the templates
+* `autoconf` and `autoheader` to generate `configure` and `config.h.in`
+
+Because of course those files are not part of the git repository.
+
+### Building YARP as part of CRuby
+
+[This script](https://github.com/ruby/ruby/blob/32e828bb4a6c65a392b2300f3bdf93008c7b6f25/tool/sync_default_gems.rb#L399-L426) imports YARP sources in CRuby.
+
+The script generates the templates when importing.
+
+`include/yarp/config.h` is replaced by `#include "ruby/config.h"`.
+It is assumed that CRuby's `./configure` is a superset of YARP's configure checks.
+
+YARP's `autotools` is not used at all in CRuby and in fact YARP's Makefile is not used either.
+Instead, CRuby's `autotools` setup is used, and `CRuby`'s Makefiles are used.
+
+### Building YARP as part of TruffleRuby
+
+[This script](https://github.com/oracle/truffleruby/blob/master/tool/import-yarp.sh) imports YARP sources in TruffleRuby.
+The script generates the templates when importing.
+It also generates `configure` and `config.h.in` (to avoid needing autotools on every machine building TruffleRuby).
+
+Then when `mx build` builds TruffleRuby and the `yarp` mx project inside, it:
+* runs `./configure`
+* runs `make`
+
+Then the `yarp bindings` mx project is built, which contains the [bindings](https://github.com/oracle/truffleruby/blob/master/src/main/c/yarp_bindings/src/yarp_bindings.c)
+and links to librubyparser.a (to avoid exporting symbols, so no conflict when installing the yarp gem).
+
+### Building YARP as part of JRuby
+
+TODO, probably similar to TruffleRuby.
```

Comment:
@enebo Could you fill this section? (can be just a comment that I copy/paste for simplicity)

Reply:
@eregon I will once I have something relevant to put here.  My playground work was manually copying files.  I probably will be doing something similar with importing of org.yarp as a tool but it doesn't exist yet (I will be restarting from main once metadata PR lands).